recursion/program/src/reduce.rs

```diff
@@ -164,15 +177,56 @@ pub fn build_reduce_program(setup: bool) -> RecursionProgram<Val> {
     }
     builder.cycle_tracker("stage-b-setup-recursion-challenger");
 
+    let expected_start_pc = builder.get(&start_pcs, zero);
+    let expected_start_shard = builder.get(&start_shards, zero);
+
     builder.range(0, num_proofs).for_each(|i, builder| {
         let proof = builder.get(&proofs, i);
         let sorted_indices = builder.get(&sorted_indices, i);
         let is_recursive = builder.get(&is_recursive_flags, i);
+
+        let shard_start_pc = builder.get(&start_pcs, i);
+        let shard_next_pc = builder.get(&next_pcs, i);
+        let shard_start_shard = builder.get(&start_shards, i);
+        let shard_next_shard = builder.get(&next_shards, i);
+
+        // Verify shard transition
+        builder.assert_felt_eq(expected_start_pc, shard_start_pc);
+        builder.assign(expected_start_pc, shard_next_pc);
+        builder.assert_felt_eq(expected_start_shard, shard_start_shard);
+        builder.assign(expected_start_shard, shard_next_shard);
+
         builder.if_eq(is_recursive, zero).then_or_else(
             // Non-recursive proof
             |builder| {
-                let shard_f = builder.get(&proof.public_values, 32);
+                let mut pv_elements = Vec::new();
+                for i in 0..PROOF_MAX_NUM_PVS {
+                    let element = builder.get(&proof.public_values, i);
+                    pv_elements.push(element);
+                }
+
+                let pv = PublicValues::<Word<Felt<_>>, Felt<_>>::from_vec(pv_elements);
+
+                // Verify witness data
+                builder.assert_felt_eq(shard_start_pc, pv.start_pc);
+                builder.assert_felt_eq(shard_next_pc, pv.next_pc);
+                builder.assert_felt_eq(shard_start_shard, pv.shard);
+                let pv_shard_plus_one: Felt<_> = builder.eval(pv.shard + one_felt);
+
+                let pv_next_pc = felt_to_var(builder, pv.next_pc);
+                builder.if_eq(pv_next_pc, zero).then_or_else(
+                    |builder| {
+                        builder.assert_felt_eq(shard_next_shard, zero_felt);
+                    },
+                    |builder| {
+                        builder.assert_felt_eq(shard_next_shard, pv_shard_plus_one);
+                    },
+                );
```

Comment:
do we need to witness/commit next_shard ? don't we only need shard?

Reply:
We don't for the first layer of verifying SP1 proofs, but I believe we need them for the higher layers.  Basically knowing what is the shard range for a recursive proof.

Reply:
To your point, the `next_shard` data is not really needed for the first layer of proof verification, so this constraint can be removed.